docs/architecture/ADR-003-Non-interactive-defaults.md

```diff
@@ -0,0 +1,691 @@
+# ADR 003: Non-interactive Defaults, Wrapped Transactions, and Subtree Root Message Inclusion Checks
+
+NOTE: Unlike normal tendermint/cosmos ADRs, this ADR isn't for deciding on whether or not we will implement non-interactive defaults. The goal of this document is to help reviewers and future readers understand what non-interactive defaults are, the considerations that went into the initial implementation, and how it differs from the original specs.
+
+The exact approach taken by the initial implementation, however, are certainly up for scrutiny. In this specific case, if there are significant changes proposed, then we might want to implement those improvements in other ADRs/PRs considering that we want to switch to a functioning version of non-interactive defaults as soon as possible.
+
+## Intro
+
+pls see the [original specs](https://github.com/celestiaorg/celestia-specs/blob/master/src/rationale/message_block_layout.md), from which this ADR paraphrases heavily.
+
+Currently, when checking for message inclusion, validators recreate the share commitment from the messages in the block and compare those with what are signed over in the `MsgPayForData` transactions also in that block. If any commitment is not found in one of the PFD transactions, or if there is a commitment that doesn't have a corresponding message, then they reject that block.
+
+While this functions as a message inclusion check, the light client has to assume that 2/3's of the voting power is honest in order to be assured that both the messages they are interested in and the rest of the messages paid for in that block are actually included. In order to have this property, we need a block (or tx? pls halp reviewers) validity rule where:
+
+> **All share commitments included in `MsgPayForData` must consist only of subtree roots of the data square.**
+
+The main issue with that requirement is that it requires the users to know the relevant subtree roots before they sign, which is problematic considering that is the block is not organized perfectly, the subtree roots will include data unknown to the user at the time of signing.
+
+To fix this, the spec outlines the “non-interactive default rules”. These involve a few additional message layout rules that allow for commitments to messages to consist entirely of sub tree roots of the data root, and for those sub tree roots to be generated only from the message itself (so that the user can sign something “non-interactively”). NOTE: MODIFIED FROM THE SPEC
+
+> **Messages must begin at a location aligned with the largest power of 2 that is not larger than the message length or k.**
+
+![Subtree root commitment](./assets/subtree-root.png "Subtree Root based commitments")
+
+> **If the messages are larger than k, then they must start on a new row.**
```

Comment:
hmmm idk, @adlerjohn is good with this sort of concise wording. I tried to stick to as close to what is written in the specs

> Messages begin at a location aligned with the largest power of 2 that is not larger than the message length or k.

the rule was mainly just modified to be first, as I felt it was more important since we're only following a portion of the other rule now

> Messages that span multiple rows must begin at the start of a row (this can occur if a message is longer than k shares or if the block producer decides to start a message partway through a row and it cannot fit).

Reply:
but we should probably be more explicit about that in the ADR come to think of it

Reply:
> Whether the power is strictly less than the length, or, less or equal.

> that is not larger than the message length or k.

I interpret this to mean not strictly less than, but less than or equal to

does that clarify this?

Reply:
Aaaah, now I get it, when calculating the largest power of 2 that is less or equal than message length:

1. We start the count from the share after the last share of the previous namespace
2. We start the new namespace at the last position of the count. For example, if message length is `7`, the largest power of 2 that is less than message length is 4. Then, we start the new namespace at the fourth share after the last share of the previous namespace. 

Thanks a lot for your help, now I get this more.

Reply:
> Then, we start the new namespace at the fourth share after the last share of the previous namespace.

just tbc, it doesn't really have that much to do with where the last namespace ended. we want to start at the next aligned power of two, because we want a single subroot to contain all of the set of 4 shares. In a square size of 8, there are exactly two spots for this, share 0, and share 4. If the largest power of two a message cannot fit on the next aligned power of two, then it must always start on the next row.

Reply:
Other questions, please, when you have time:

1. In your previous response:
> In a square size of 8, there are exactly two spots for this, share 0, and share 4

You mean share 0 and share 3, right? Because the message will be aligned with the power of 2?

2) If `ns = 1` had 7 shares. Where would it start?
- The beginning of a new row
- The 4th position in a new row
- Where it is now?

3) if the `k = 8`, then the largest power of 2 that are possible are: 1, 2, 4, and 8. 
If a message length is 4. Then, the possible start of message are 1, 2 and 4. Then, the block producer checks if that collides with a previous namespace, then it does that on a new row, right?
In the above diagram, `ns = 4`  is still a mystery to me.

Reply:
I have other questions, please, when you have time :smile_cat:  :

1. In your previous response:
> In a square size of 8, there are exactly two spots for this, share 0, and share 4

You mean share 0 and share 3, right? Because the message will be aligned with the power of 2?

2) If `ns = 1` had 7 shares. Where would it start?
- The beginning of a new row
- The 4th position in a new row
- The same position it is starting now

3) if the `k = 8`, then the largest power of 2 that are possible are: 1, 2, 4, and 8. 
If a message length is 4. Then, the possible start of message are 1, 2 and 4. Then, the block producer checks if that collides with a previous namespace, then it does that on a new row, right?
In the above diagram, `ns = 4`  is still a mystery to me.

And, for the clients, if I understand right, all they need to do is decompose the message to a  merkle mountain range structure, that is bound by the row size (which is always a power of 2), and sign all the subtree roots commitments, for all row sizes.

Reply:
> You mean share 0 and share 3, right? Because the message will be aligned with the power of 2?

we actually want 4 in this case, as we want to start at the next aligned power of two, and we would want to start on the next row (5th position).

> f ns = 1 had 7 shares. Where would it start?

if there was room on the row for 4 shares, the largest power of two less than or equal to the size of _the message_, then it would start there. If not, then it would start on the next row.

> if the k = 8, then the largest power of 2 that are possible are: 1, 2, 4, and 8.

yeah!

> If a message length is 4. Then, the possible start of message are 1, 2 and 4.

then we have to use the largest, 4. it has to start at a share index that is divisible by 4